### PR TITLE
Biomedical Schema Draft

### PR DESCRIPTION
**Pull request by jason-fries**
- Please comment on the candidate task schemas. 
- Use PR review to add your thoughts and propose changes.

### CODE REVIEW

File: examples/bioasq9b.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            name="original",
+            version=VERSION,
+            description="Original BioASQ9 source schema.",
+        ),
+        datasets.BuilderConfig(
+            name="bigbio",
+            version=VERSION,
+            description="Simplified schema for BigScience-Biomedical QA tasks",
+        ),
+    ]
```
Comment (galtay) [inline examples/bioasq9b.py:96]: I like the `original` and `bigbio` config names ... we might want to do something like `bigbio-schema` (e.g. bigbio-kb or bigbio-text-pairs) as some datasets support multiple tasks. Specifically I'm thinking of things like the MuchMore dataset which can be a translation dataset or an NER dataset. `bigbio-<schema>` instead of `bigbio-task` or just `bigbio` handles this well i think

File: examples/bioasq9b.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                            "text": datasets.Value("string"),
+                            "beginSection": datasets.Value("string"),
+                            "endSection": datasets.Value("string"),
+                            "document": datasets.Value("string"),
+                        }
+                    ),
+                }
+            )
+        # simplified scheam for QA tasks
+        elif self.config.name == "bigbio":
```
Comment (galtay) [inline examples/bioasq9b.py:133]: e.g. `bigbio-qa` here instead of `bigbio` ... this would also allow some nice discovery by investigating the configs of the datasets

File: examples/bioasq9b.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                name=datasets.Split.TEST,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(
+                        data_dir, "BioASQ-training9b/training9b.json"
+                    ),
+                    "split": "test",
+                },
+            ),
+        ]
```
Comment (galtay) [inline examples/bioasq9b.py:193]: we should probably leave some instructions in the hackathon README about what to do if the dataset does not contain a canonical train/val/test split. looks like these files are the same for train and test ... I think we should default to just a `train` set if there is no canonical split.

File: examples/bioasq9b.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    ):
+        """Yields examples as (key, example) tuples."""
+        # This method handles input defined in _split_generators to yield (key, example) tuples from the dataset.
+        # The `key` is here for legacy reason (tfds) and is not important in itself.
+
+        with open(filepath, encoding="utf-8") as file:
+            data = json.load(file)
+            for record in enumerate(data):
+                if self.config.name == "original":
+                    yield record["id"], {
```
Comment (galtay) [inline examples/bioasq9b.py:208]: I might be confused, but I think `record` here will be a tuple with the first element i = 0 to len(data)-1 and the second element being data[i]  .. so `record["id"]` seems like it would throw an error. I think maybe we want something like, 
```
for _id, record in enumerate(data):
    yield _id, {"id": record["id"], ...}
```

File: schemas/entailment.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+"""
+import datasets
+
+
+features = datasets.Features(
+    {
+        "id": datasets.Value("string"),
+        "premise": datasets.Value("string"),
+        "hypothesis": datasets.Value("string"),
+        "label": datasets.Value("string"),
```
Comment (galtay) [inline schemas/entailment.py:12]: add even more guide rails and make this 
`"label": datasets.ClassLabel(names=["entailment", "contradiction", "neutral"])`
or too strict?

File: schemas/kb.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+and a flat hierarchy of document passages.
+
+"""
+import datasets
+
+
+features = datasets.Features(
+    {
+        "id": datasets.Value("string"),
+        "document_id": datasets.Value("string"),
```
Comment (galtay) [inline schemas/kb.py:23]: I think it would be nice to have a top level `text` field here that any offset would work in ... for example
```
text = sample["text"]
ii = sample["passages"][n]["offsets"][0][0]
ff = sample["passages"][n]["offsets"][0][1]
assert(text[ii:ff] = sample["passages"][n]["text"])
# same for entities 
```

Otherwise, if I'm looking for the context of an entity I have to search through each passage to discover if the entity offset is within the offset range of that passage. This could be as straightforward as, 
```
sample["text"] = " ".join([passage['text'] for passage in passages])
```

It definitely leads to some duplicate information but I think its worth it for reasonable sized datasets ... it would also give the unit tests something to check.

File: task_schemas.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+- Named entity recognition (NER)
+- Named entity disambiguation/normalization/linking (NED)
+- Event extraction
+- Relation extraction (RE)
+- Coreference resolution
+
+```
+{
+    "document_id": "XXXXXX",
```
Comment (galtay) [inline task_schemas.md:31]: we have `id` and `document_id` in the `schemas/kb.py`  maybe we can go into some detail here about how they should be used?

File: task_schemas.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+
+### Passages
+
+Passages capture document structure, such as the title and abstact sections of a PubMed abstract. 
+
+```
+{
+    "id": 0,
+    "document_id": "227508",
```
Comment (galtay) [inline task_schemas.md:60]: this makes the distinction between `id` and `document_id` a bit more clear .... do we want `id` to be something like "a unique hashable item in the dataset"?

File: examples/bioasq9b.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
+# coding=utf-8
+# Copyright 2020 The HuggingFace Datasets Authors and the current dataset script contributor.
```
Comment (galtay) [inline examples/bioasq9b.py:2]: do we want to have people add their name here in place of `the current dataset script contributor`? or have people append their name if they make a PR?

File: schemas/kb.py (hunk 6 above)
Comment (galtay) [reply schemas/kb.py:23]: coming back to this (sorry for the long comments!) ... I now see that my suggestion would be particularly silly for samples with one passage (then the top level text would just be an exact copy of the `passages[0]['text']`  ... I withdraw my top level text suggestion :)

File: task_schemas.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+
+**Schema Notes**
+
+- `id` must be unique string within dataset and across parent category (e.g., entities, passages)
+- `offsets` contains absolute character offsets per span
+- `offsets` and `text` are lists to support discontinous spans. 
+- `normalized` may contain 1 or more normalized links to database entity identifiers.
+- `passages` captures document structure such as named sections.
+- `entities`,`events`,`coreferences`,`relations` may be empty fields based on dataset and specific task.
```
Comment (galtay) [inline task_schemas.md:49]: we might even go as far as to leave these fields out if they are not included in the dataset?

File: schemas/kb.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                "type": datasets.Value("string"),
+                "text": datasets.Value("string"),
+                "offsets": datasets.Sequence([datasets.Value("int32")]),
+            }
+        ),
+        "entities": datasets.Sequence(
+            {
+                "id": datasets.Value("string"),
+                "offsets": datasets.Sequence([datasets.Value("int32")]),
+                "text": datasets.Sequence([datasets.Value("string")]),
```
Comment (galtay) [inline schemas/kb.py:36]: I think this should be just `Sequence(Value("string"))` not a nested iterable

File: schemas/kb.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+features = datasets.Features(
+    {
+        "id": datasets.Value("string"),
+        "document_id": datasets.Value("string"),
+        "passages": datasets.Sequence(
+            {
+                "id": datasets.Value("string"),
+                "type": datasets.Value("string"),
+                "text": datasets.Value("string"),
+                "offsets": datasets.Sequence([datasets.Value("int32")]),
```
Comment (galtay) [inline schemas/kb.py:29]: if we're not allowing for disjoint passages I think we should have just `Sequence(Value("int32"))` here  (OR if we are, then make the `text` a sequence as well). I think we don't need disjoint passages.

File: examples/bioasq9b.py (hunk 3 above)
Comment (jason-fries) [reply examples/bioasq9b.py:193]: This is just an oversight in this case -- there is a canonical test set. More generally, I think we should be in a setting where there is always a train split defined and (hopefully) a public test or validation set. If not, defaulting to `train` is good with me

File: schemas/entailment.py (hunk 5 above)
Comment (jason-fries) [reply schemas/entailment.py:12]: Yeah, I'm torn. Let's leave this unrestricted for now and see if it becomes a problem and we can flag this for fixing in a post-hackathon cleanup pass.

File: schemas/kb.py (hunk 11 above)
Comment (jason-fries) [reply schemas/kb.py:36]: This is to match the discontinuous offset definition. The ugliness of supporting discontinuous spans necessitates this unfortunately.

Comment (galtay) [reply schemas/kb.py:36]: Even for the discontinuous case, a nested set of offsets `[(0,4), (9,12)]` leads to one less level of nesting for the text right? `["text1", "text2"]`
